src/assets/Generator.Shared/NextLinkOperationImplementation.cs

```diff
@@ -215,34 +197,63 @@ internal static bool TryGetApiVersion(Uri startRequestUri, out ReadOnlySpan<char
         /// <summary>
         /// This function is used to get the final request uri after the lro has completed.
         /// </summary>
-        private string? GetFinalUri()
+        private string? GetFinalUri(Response response)
         {
-            // Set final uri as null if the response header doesn't contain "Operation-Location" or "Azure-AsyncOperation".
+            // Set final uri as null if the response for initial request doesn't contain header "Operation-Location" or "Azure-AsyncOperation".
             if (_headerSource is not (HeaderSource.OperationLocation or HeaderSource.AzureAsyncOperation))
             {
                 return null;
             }
 
-            // Set final uri as null if original request is a delete method.
+            // Handle final-state-via options: https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-long-running-operation-options
+            switch (_finalStateVia)
+            {
+                case OperationFinalStateVia.Location when _originalResponseHasLocation:
+                    return _lastKnownLocation;
+                case OperationFinalStateVia.OperationLocation or OperationFinalStateVia.AzureAsyncOperation:
+                    return null;
+                case OperationFinalStateVia.OriginalUri:
+                    return _startRequestUri.AbsoluteUri;
+            }
+
+            // Set final uri as null if initial request is a delete method.
             if (_requestMethod == RequestMethod.Delete)
             {
                 return null;
             }
 
-            // Set final uri as original request's uri if one of these requirements are met:
-            // 1.Original request is a put method;
-            // 2.Original request is a management plane patch method. For data plane patch method, the final uri will be determinned by the original response header and "_finalStateVia";
-            // 3.Original response header contains "Location" and FinalStateVia is configured to OriginalUri.
-            if (_requestMethod == RequestMethod.Put || _requestMethod == RequestMethod.Patch && _startRequestUri.Scheme == "https" && _startRequestUri.Host == "management.azure.com" || _originalResponseHasLocation && _finalStateVia == OperationFinalStateVia.OriginalUri)
+            // If body contains resourceLocation, use it
```

Comment:
This is still overly expensive - to parse JSON, which could be large - just to skip past this. How can we short-circuit this? If I set the `final-state-via` to `original-url` - or if it was set to anything in particular that doesn't indicate a body - could we skip this?

Reply:
If `final-state-via` is set to anything, then it will be handled without parsing the body. 
@lmazuel, is it possible that in case of POST we try to parse payload without any good cause?